content/6_Advanced/SCC.mdx

```diff
@@ -258,24 +258,231 @@ int main() {
+- For each variable $x$, we compare $comp[x]$ and $comp[\lnot x]$. If $comp[x] = comp[\lnot x]$, it 	indicates that both $x$ and $\lnot x$ belong to same SCCs, implying a contradiction.
+- In such cases, we return $false$ to indicate that no valid assignment satisfies the 2-SAT problem.
+
+Below is the implementation of the solution for the 2-SAT problem, utilizing the graph of implications $adj$ and its transpose $adj^T$, where vertices $2k$ and $2k+1$ correspond to variable $k$ and its negation respectively.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+struct TwoSatSolver {
+	int n_vars;
+	int n_vertices;
+	vector<vector<int>> adj, adj_t;
+	vector<bool> used;
+	vector<int> order, comp;
+	vector<bool> assignment;
+
+	TwoSatSolver(int _n_vars)
+	    : n_vars(_n_vars), n_vertices(2 * n_vars), adj(n_vertices),
+	      adj_t(n_vertices), used(n_vertices), order(), comp(n_vertices, -1),
+	      assignment(n_vars) {
+		order.reserve(n_vertices);
+	}
+	void dfs1(int v) {
+		used[v] = true;
+		for (int u : adj[v]) {
+			if (!used[u]) dfs1(u);
+		}
+		order.push_back(v);
+	}
+
+	void dfs2(int v, int cl) {
+		comp[v] = cl;
+		for (int u : adj_t[v]) {
+			if (comp[u] == -1) dfs2(u, cl);
+		}
+	}
+
+	bool solve_2SAT() {
+		order.clear();
+		used.assign(n_vertices, false);
+		for (int i = 0; i < n_vertices; ++i) {
+			if (!used[i]) dfs1(i);
+		}
+
+		comp.assign(n_vertices, -1);
+		for (int i = 0, j = 0; i < n_vertices; ++i) {
+			int v = order[n_vertices - i - 1];
+			if (comp[v] == -1) dfs2(v, j++);
+		}
+
+		assignment.assign(n_vars, false);
+		for (int i = 0; i < n_vertices; i += 2) {
+			if (comp[i] == comp[i + 1]) return false;
+			assignment[i / 2] = comp[i] > comp[i + 1];
+		}
+		return true;
+	}
+
+	void add_disjunction(int a, bool na, int b, bool nb) {
+		// na and nb signify whether a and b are to be negated
+		a = 2 * a ^ na;
+		b = 2 * b ^ nb;
+		int neg_a = a ^ 1;
+		int neg_b = b ^ 1;
+		adj[neg_a].push_back(b);
+		adj[neg_b].push_back(a);
+		adj_t[b].push_back(neg_a);
+		adj_t[a].push_back(neg_b);
+	}
+
+	static void example_usage() {
+		TwoSatSolver solver(3);                      // a, b, c
+		solver.add_disjunction(0, false, 1, true);   //     a  v  not b
+		solver.add_disjunction(0, true, 1, true);    // not a  v  not b
+		solver.add_disjunction(1, false, 2, false);  //     b  v      c
+		solver.add_disjunction(0, false, 0, false);  //     a  v      a
+		assert(solver.solve_2SAT() == true);
+		auto expected = vector<bool>(True, False, True);
+		assert(solver.assignment == expected);
+	}
+};
+```
+</CPPSection>
+</LanguageSection>
+<FocusProblem problem="satEx" />
+
+It is a straightforward 2SAT problem. Each topping corresponds to a variable. When a topping $x$ is preferred with '+', we include $x$ in our clause. For preference '-', we include $\lnot x$ in the clause.
+
+## Solution
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct TwoSatSolver {
```

Comment:
wrap in codesnip

Reply:
and what is that?

Reply:
https://usaco.guide/general/working-mdx
should be somewhere in here

Reply:
understood it

Reply:
done